File: algorithms/labs/lab01/MaxHeap.java
// Heap code in Java
// Pramod Nepal

import java.util.Scanner;

public class MaxHeap {
    private int max_size;           //max size of heap, this is included just to explain that size is not maximum size, but is current heap size.
    private int size;               //size of current heap
    private String[] h;             //array of strings that the heap will use

    public MaxHeap(int max_size) 
    {
        size = 1;
        this.max_size = max_size + 1;
        h = new String[this.max_size];       //we are not using the 0th position of the array, so we need one additional memory spot
    }
    
    public void Print()
    {
        for(int i = 1; i <= size-1; i++) {
            System.out.print(h[i] + " ");
        }
    }
    
    public void Insert(String item)
    {           
        h[size] = item;         
        FixHeapIter(size);
        size++;
    }   
    
    
    public void swap(int first, int second)
    {
        String tmp = h[first];
        h[first] = h[second];
        h[second] = tmp;
    }
    
    public void FixHeapIter(int n) 
    {
        
        if (n > 1) { 
        
            int parent =  n / 2;
        
        
            int compare = h[n].compareTo(h[parent]);
        
            if (compare > 0) {
                swap(n, parent);
                FixHeapIter(parent);
            }
        
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int size;

        System.out.print("Number of chars: ");
        size = scanner.nextInt();

        MaxHeap maxHeap = new MaxHeap(size);

        for (int i = 1; i <= size; i++) {
            String ch = scanner.next();
            maxHeap.Insert(ch);
        }

        maxHeap.Print();
    }
}

File: algorithms/labs/lab01/maxheap.cpp
#include <iostream>

using namespace std;

int max_size;
int size;
char *h;

void Print()
{
    for (int i = 1; i < size; i++)
        cout << h[i];
}

int ParentIndex(int n)
{
    return (n)/2;   
}

void Swap(int n, int p)
{
    char tmp = h[n];
    h[n] = h[p];
    h[p] = tmp;
}

void HeapSort(int n)
{
    if (n > 1) {
        int p = ParentIndex(n);
        if (h[n] > h[p]) {
            Swap(n, p);
            HeapSort(p);
        }
    }
}

void Insert(char c)
{
   h[size] = c;
   HeapSort(size); 
   size++;
}


int main(int argc, char *argv[])
{
    size = 1;
    cout << "Number of items: ";
    cin >> max_size;
    max_size = max_size + 1;
    char inputChar;
    h = new char[max_size];

    for(int i = 1; i <= max_size-1; i++) {
        cin >> inputChar;
        Insert(inputChar);
    }

    Print();

    delete [] h;
}
